src/unix/signal.c

```diff
@@ -318,7 +318,11 @@ void deliver_signal_to_process(process p, struct siginfo *info)
 
     /* If a thread is set as runnable and can handle this signal, just return. */
     thread t, can_wake = 0;
-    vector_foreach(p->threads, t) {
+    vector v = allocate_vector(heap_general(get_kernel_heaps()), 8);
+    spin_lock(&p->threads_lock);
+    threads_to_vector(p, v);
+    spin_unlock(&p->threads_lock);
+    vector_foreach(v, t) {
```

Comment:
I suppose the transfer to vector could have performance implications if an instance contains a lot of threads and regularly accepts signals delivered at the process level (e.g. timer_create / SIGEV_SIGNAL).

Could there be a way to embed this thread search in an rbtree_traverse directly? Most of this is just checking flags and should be safe to do while threads_lock is held. The call to signalfd_dispatch could move outside the loop. (I was trying to recall why there wasn't a break after can_wake is assigned - I think it's to keep looking for a thread that is already runnable.)

Reply:
I moved the search into the traversal, but kept the signalfd_dispatch out. This changes behavior very slightly in that before, if applying the dispatch call made the thread runnable, it could jump out of the loop instead of continuing the search, and this can't happen now with signalfd_dispatch out of the loop.

I do think the new code makes more sense, avoids an alloc/dealloc and so is more performant, and I don't think the missing short circuit is a big deal compared to that.